sxiv.1
.TH SXIV 1 sxiv\-VERSION
.SH NAME
sxiv \- Simple X Image Viewer
.SH SYNOPSIS
.B sxiv
.RB [ \-abcfhiopqrtvZ ]
.RB [ \-A
.IR FRAMERATE ]
.RB [ \-e
.IR WID ]
.RB [ \-G
.IR GAMMA ]
.RB [ \-g
.IR GEOMETRY ]
.RB [ \-N
.IR NAME ]
.RB [ \-T
.IR TITLE ]
.RB [ \-n
.IR NUM ]
.RB [ \-S
.IR DELAY ]
.RB [ \-s
.IR MODE ]
.RB [ \-z
.IR ZOOM ]
.IR FILE ...
.SH DESCRIPTION
sxiv is a simple image viewer for X.
.P
It has two modes of operation: image and thumbnail mode. The default is image
mode, in which only the current image is shown. In thumbnail mode a grid of 
small previews is displayed, making it easy to choose an image to open.
.P
Please note, that the fullscreen mode requires an EWMH/NetWM compliant window
manager.
.SH OPTIONS
.TP
.BI "\-A " FRAMERATE
Play animations with a constant frame rate set to
.IR FRAMERATE .
.TP
.B \-a
Play animations of multi-frame images.
.TP
.B \-b
Do not show info bar on bottom of window.
.TP
.B \-c
Remove all orphaned cache files from the thumbnail cache directory and exit.
.TP
.BI "\-e " WID
Embed sxiv's window into window whose ID is
.IR WID .
.TP
.B \-f
Start in fullscreen mode.
.TP
.BI "\-G " GAMMA
Set image gamma to GAMMA (-32..32).
.TP
.BI "\-g " GEOMETRY
Set window position and size. See section GEOMETRY SPECIFICATIONS of X(7) for
more information on GEOMETRY argument.
.TP
.BI "\-N " NAME
Set the resource name of sxiv's X window to NAME.
.TP
.BI "\-T " TITLE
Set the title of sxiv's X window to TITLE.
.TP
.BI "\-n " NUM
Start at picture number NUM.
.TP
.B \-h
Print brief usage information to standard output and exit.
.TP
.B \-i
Read names of files to open from standard input. Also done if FILE is `-'.
.TP
.B \-o
Write list of all marked files to standard output when quitting. In combination
with
.B \-i
sxiv can be used as a visual filter/pipe.
.TP
.B \-p
Enable private mode, in which sxiv does not write any cache or temporary files.
.TP
.B \-q
Be quiet, disable warnings to standard error stream.
.TP
.B \-r
Search the given directories recursively for images to view.
.TP
.BI "\-S " DELAY
Start in slideshow mode. Set the delay between images to
.I DELAY
seconds.
.I DELAY
may be a floating point number.
.TP
.BI "\-s " MODE
Set scale mode according to MODE character. Supported modes are: [d]own, 
[f]it, [w]idth, [h]eight.
.TP
.B \-t
Start in thumbnail mode.
.TP
.B \-v
Print version information to standard output and exit.
.TP
.B \-Z
The same as `\-z 100'.
.TP
.BI "\-z " ZOOM
Set zoom level to ZOOM percent.
.SH KEYBOARD COMMANDS
.SS General
The following keyboard commands are available in both image and thumbnail mode:
.TP
.BR 0 \- 9
Prefix the next command with a number (denoted via
.IR count ).
.TP
.B q
Quit sxiv.
.TP
.B Return
Switch to thumbnail mode / open selected image in image mode.
.TP
.B f
Toggle fullscreen mode.
.TP
.B b
Toggle visibility of info bar on bottom of window.
.TP
.B Ctrl-x
Send the next key to the external key-handler. See section EXTERNAL KEY HANDLER
for more information.
.TP
.B g
Go to the first image.
.TP
.B G
Go to the last image, or image number
.IR count .
.TP
.B r
Reload image.
.TP
.B D
Remove current image from file list and go to next image.
.TP
.BR Ctrl-h ", " Ctrl-Left
Scroll left one screen width.
.TP
.BR Ctrl-j ", " Ctrl-Down
Scroll down one screen height.
.TP
.BR Ctrl-k ", " Ctrl-Up
Scroll up one screen height.
.TP
.BR Ctrl-l ", " Ctrl-Right
Scroll right one screen width.
.TP
.BR +
Zoom in.
.TP
.B \-
Zoom out.
.TP
.B m
Mark/unmark the current image.
.TP
.B M
Reverse all image marks.
.TP
.B Ctrl-M
Repeat last mark action on all images from the last marked/unmarked up to the
current one.
.TP
.B Ctrl-m
Remove all image marks.
.TP
.B N
Go
.I count
marked images forward.
.TP
.B P
Go
.I count
marked images backward.
.TP
.B {
Decrease gamma correction by
.I count
steps.
.TP
.B }
Increase gamma correction by
.I count
steps.
.TP
.B Ctrl-g
Reset gamma correction.
.SS Thumbnail mode
The following keyboard commands are only available in thumbnail mode:
.TP
.BR h ", " Left
Move selection left
.I count
times.
.TP
.BR j ", " Down
Move selection down
.I count
times.
.TP
.BR k ", " Up
Move selection up
.I count
times.
.TP
.BR l ", " Right
Move selection right
.I count
times.
.TP
.B R
Reload all thumbnails.
.SS Image mode
The following keyboard commands are only available in image mode:
.TP
Navigate image list:
.TP
.BR n ", " Space
Go
.I count
images forward.
.TP
.BR p ", " Backspace
Go
.I count
images backward.
.TP
.B [
Go
.I count
* 10 images backward.
.TP
.B ]
Go
.I count
* 10 images forward.
.TP
Handle multi-frame images:
.TP
.B Ctrl-n
Go
.I count
frames of a multi-frame image forward.
.TP
.B Ctrl-p
Go
.I count
frames of a multi-frame image backward.
.TP
.B Ctrl-Space
Play/stop animations of multi-frame images.
.TP
Panning:
.TP
.BR h ", " Left
Scroll image 1/5 of window width or
.I count
pixel left.
.TP
.BR j ", " Down
Scroll image 1/5 of window height or
.I count
pixel down.
.TP
.BR k ", " Up
Scroll image 1/5 of window height or
.I count
pixel up.
.TP
.BR l ", " Right
Scroll image 1/5 of window width or
.I count
pixel right.
.TP
.B H
Scroll to left image edge.
.TP
.B J
Scroll to bottom image edge.
.TP
.B K
Scroll to top image edge.
.TP
.B L
Scroll to right image edge.
.TP
Zooming:
.TP
.B =
Set zoom level to 100%, or
.IR count %.
.TP
.B w
Set zoom level to 100%, but fit large images into window.
.TP
.B W
Fit image to window.
.TP
.B e
Fit image to window width.
.TP
.B E
Fit image to window height.
.TP
Rotation:
.TP
.B <
Rotate image counter-clockwise by 90 degrees.
.TP
.B >
Rotate image clockwise by 90 degrees.
.TP
.B ?
Rotate image by 180 degrees.
.TP
Flipping:
.TP
.B |
Flip image horizontally.
.TP
.B _
Flip image vertically.
.TP
Miscellaneous:
.TP
.B a
Toggle anti-aliasing.
.TP
.B A
Toggle visibility of alpha-channel, i.e. image transparency.
.TP
.B s
Toggle slideshow mode and/or set the delay between images to
.I count
seconds.
.SH MOUSE COMMANDS
The following mouse mappings are available in image mode:
.TP
General:
.TP
.B Button3
Switch to thumbnail mode.
.TP
Navigate image list:
.TP
.B Button1
Go to the next image if the mouse cursor is in the right part of the window or
to the previous image if it is in the left part.
.TP
Panning:
.TP
.B Button2
Pan the image according to the mouse cursor position in the window while
keeping this button pressed down.
.TP
Zooming:
.TP
.B ScrollUp
Zoom in.
.TP
.B ScrollDown
Zoom out.
.SH CONFIGURATION
The following X resources are supported:
.TP
.B background
Color of the window background and bar foreground
.TP
.B foreground
Color of the window foreground and bar background
.TP
.B font
Name of Xft bar font
.TP
Please see xrdb(1) on how to change them.
.SH STATUS BAR
The information displayed on the left side of the status bar can be replaced
with the output of a user-provided script, which is called by sxiv whenever an
image gets loaded. The path of this script is
.I $XDG_CONFIG_HOME/sxiv/exec/image-info
and the arguments given to it are: 1) path to image file, 2) image width,
3) image height.
.P
There is also an example script installed together with sxiv as
.IR PREFIX/share/sxiv/exec/image-info .
.SH EXTERNAL KEY HANDLER
Additional external keyboard commands can be defined using a handler program
located in
.IR $XDG_CONFIG_HOME/sxiv/exec/key-handler .
The handler is invoked by pressing
.BR Ctrl-x .
The next key combo is passed as its first argument. Passed via stdin are the
images to act upon, one path per line: all marked images, if in thumbnail mode
and at least one image has been marked, otherwise the current image.
sxiv(1) will block until the handler terminates. It then checks which images
have been modified and reloads them.

The key combo argument has the following form: "[C-][M-][S-]KEY",
where C/M/S indicate Ctrl/Meta(Alt)/Shift modifier states and KEY is the X
keysym as listed in /usr/include/X11/keysymdef.h without the "XK_" prefix.

There is also an example script installed together with sxiv as
.IR PREFIX/share/sxiv/exec/key-handler .
.SH THUMBNAIL CACHING
sxiv stores all thumbnails under
.IR $XDG_CACHE_HOME/sxiv/ .
.P
Use the command line option
.I \-c
to remove all orphaned cache files. Additionally, run the following command
afterwards inside the cache directory to remove empty subdirectories:
.P
.RS
find . \-depth \-type d \-empty ! \-name '.' \-exec rmdir {} \\;
.RE
.SH AUTHOR
.EX
Bert Muennich   <xyb3rt at gmail.com>
.EE
.SH CONTRIBUTORS
.EX
Bastien Dejean  <nihilhill at gmail.com>
Dave Reisner    <d at falconindy.com>
Fung SzeTat     <sthorde at gmail.com>
Max Voit        <mvdev at with-eyes.net>
.EE
.SH HOMEPAGE
.EX
https://github.com/xyb3rt/sxiv
.EE
.SH SEE ALSO
.BR X (7),
.BR xrdb (1)


ChangeLog
commit 5eb082a1eb9be173aeeb9fac7d14f06c1b930b73
Author: Victor Frankenstein <victor@frankenstein>
Date:   Sat Nov 23 07:17:04 2013 +0400

    forgot to add LICENSE, added it into COPYING

commit bf6e5986698222ff57f467aabe525ad7102ec714
Author: Victor Frankenstein <victor@frankenstein>
Date:   Sat Nov 23 07:10:23 2013 +0400

    commit 4780ec5463ee49a30d55c1c63d9947eee69e9d66 ('Close info file descriptor after reading')

commit 4f3689ebb96d32776cb84f58cc9d83f34d63eb08
Author: Victor Frankenstein <victor@frankenstein>
Date:   Sat Nov 23 07:03:39 2013 +0400

    added a couple of header checks

commit 8d5727402c203aeabd43a69d1e72849a31224a26
Author: Victor Frankenstein <victor@frankenstein>
Date:   Sat Nov 23 06:55:45 2013 +0400

    restored old README.md

commit 843596fb09cdc18b478ba135fdd0e479127a7cfe
Author: Victor Frankenstein <victor@frankenstein>
Date:   Sat Nov 23 06:54:11 2013 +0400

    made a distro

commit 309639888eb538edaba1b5e451bd092cda046017
Author: Victor Frankenstein <victor@frankenstein>
Date:   Thu Nov 21 00:31:57 2013 +0400

    there's no exit from slideshow now, but it works somehow

commit cbe9c90dd6910d30ae4db8cd94938bb094ceb585
Author: Victor Frankenstein <victor@frankenstein>
Date:   Wed Nov 20 23:57:44 2013 +0400

    tried to add a slideshow, but it hangs on redraw (imlib_render_image_part_on_drawable_at_size)

commit eeb531635be909c5f5df0b326f0965168d98adf8
Author: Victor Frankenstein <victor@frankenstein>
Date:   Wed Nov 20 21:15:07 2013 +0400

    added p-get-file-index and p-get-file-count functions

commit 0451e069a6f3631eb9a24756be9e6fe068085ed4
Author: Victor Frankenstein <victor@frankenstein>
Date:   Wed Nov 20 21:02:48 2013 +0400

    added it-add-image function

commit 9ba734fcc63bafe9e413625cddf10c978a298cc9
Author: Victor Frankenstein <victor@frankenstein>
Date:   Wed Nov 20 19:16:47 2013 +0400

    removed another useless function

commit 4372d9e8cd03dbdd4f0af40985dab0047a181679
Author: Victor Frankenstein <victor@frankenstein>
Date:   Wed Nov 20 19:13:53 2013 +0400

    fix

commit 98db6d5f9bdca407fee57bed4771b6ca9301bdfb
Author: Victor Frankenstein <victor@frankenstein>
Date:   Wed Nov 20 19:09:34 2013 +0400

    made i-navigate work in thumb mode

commit 73d44c91f4a3493c66a023a4278f2389ee86da59
Author: Victor Frankenstein <victor@frankenstein>
Date:   Wed Nov 20 14:34:26 2013 +0400

    removing most useless stuff

commit 4483cce2dab1e9c9f41286aaf1e2557dec8a0bde
Author: Victor Frankenstein <victor@frankenstein>
Date:   Wed Nov 20 14:00:13 2013 +0400

    fixed output

commit 519787803255fa6d9fca9f09f8c9683f9dc0586e
Author: Victor Frankenstein <victor@frankenstein>
Date:   Wed Nov 20 13:51:38 2013 +0400

    added scheme repl support (C-e) and output of input

commit bba28cccea7d87ada12940e323b4d5cfbb714eeb
Author: Victor Frankenstein <victor@frankenstein>
Date:   Wed Nov 20 12:17:36 2013 +0400

    added mouse support

commit 92d61b12fbb4f78d27ce6c72bef7a03b638d21c5
Author: Victor Frankenstein <victor@frankenstein>
Date:   Wed Nov 20 11:30:56 2013 +0400

    added more hotkeys

commit 7df6000e8985ec31eac8cefa3fc7b5e75d624ce8
Author: Victor Frankenstein <victor@frankenstein>
Date:   Wed Nov 20 08:02:42 2013 +0400

    added some guile support, work in progress

commit 1b089bc2fa36fdbd69af25a15ec5ee3e8963b80d
Author: Bert Münnich <ber.t@posteo.de>
Date:   Sat Oct 5 17:37:00 2013 +0200

    Seek over EXIF APP0 frame instead of reading and discarding it

commit 5dea695c719dc2ee61e7677d553f16c1c4c3ac52
Author: Jacek Naglak <jnaglak@tlen.pl>
Date:   Tue Sep 24 22:14:47 2013 +0200

    Fixed image orientation if a JFIF APP0 segment is present in a JPEG header.

commit 745eacbcf5404cd0b597ac4d986e819c3e03e786
Author: Bert Münnich <be.muennich@gmail.com>
Date:   Sat Sep 7 17:59:58 2013 +0200

    Updated unstable version number

commit b98a48576e380e6f3f953b3d01b25172449e14d7
Author: Peter Hofmann <scm@uninformativ.de>
Date:   Sat Sep 7 17:06:47 2013 +0200

    Warp pointer on first/last pixel
    
    Previously, the pointer would get warped when it had left the window.
    This doesn't work when sxiv is run in fullscreen mode (or similar -- see
    dwm's monocle mode): The pointer can never leave the window in such
    situations.
    
    To fix this, warp the pointer when it's on the first/last pixel. To
    avoid an endless warping loop, the new position must be at least one
    pixel away from the next warp position.

commit 26a624a543f2e723550996d76d90ccbb7b954aeb
Author: Bert Münnich <be.muennich@gmail.com>
Date:   Thu Aug 22 12:59:05 2013 +0200

    Added options for anti-alias & alpha layer coloring to config.def.h
    
    Also removed now obsolete -p command line option; fixes issue #98

commit fb6e4bdd980f72082a264894c12ed7e0650a5bf9
Author: Bert Münnich <be.muennich@gmail.com>
Date:   Thu Aug 22 09:44:14 2013 +0200

    Only print marked files when -o is given
    
    Fallback to print all files, if no images are marked, was kind of
    irritating.

commit 7d878bd16df1abffde780e195144b845dc133166
Author: Bert Münnich <be.muennich@gmail.com>
Date:   Sat Aug 10 21:18:53 2013 +0200

    Added file marks; fixes issue #94
    
    - Command it_toggle_image_mark (bound to 'm') toggles mark of current
      image
    - Command it_navigate_marked (bound to 'N'/'P') can be used to go to
      the next/previous marked image
    - When option -o is given, all marked files get printed

commit 84d77b173288f838ffdbe2cd12061cf7a725f185
Author: Bert Münnich <be.muennich@gmail.com>
Date:   Sat Aug 10 15:59:45 2013 +0200

    Updated unstable version number

commit c6bc6069d6b6100cd1cc9af4822bf71b43fbd6d5
Author: Bert Münnich <be.muennich@gmail.com>
Date:   Sat Aug 10 15:58:16 2013 +0200

    Changed key bindings for flipping to bar & underscore

commit 56142bd3bafedf51882b368ab08a580ac98f2619
Author: Bert Münnich <be.muennich@gmail.com>
Date:   Sat Aug 10 15:55:18 2013 +0200

    Refactored merged rotation code


conf/sxiv.scm
(use-modules (ice-9 match)
             (ice-9 rdelim)
             (ice-9 regex)
             (srfi srfi-1))

(define constants
  '((left 0)
    (right 1)
    (up 2)
    (down 3)
    (scale-down 0)
    (scale-fit 1)
    (scale-width 2)
    (scale-height 3)
    (scale-zoom 4)
    (degree-90 1)
    (degree-180 2)
    (degree-270 3)
    (flip-horizontal 0)
    (flip-vertical 1)))

(define-syntax const
  (syntax-rules ()
    ((_ name)
     (cadr (assoc 'name constants)))))

(it-switch-mode)

(define *input* "")
(define *slideshow*)

(define (start-slideshow rand time)
  (define (slideshow-proc n cnt time)
    (display `(sleeping ,time))
    (sleep time)
    (yield)
    (let ((next (if n
                    (if (< (+ n 1) cnt)
                        (+ n 1)
                        0)
                    (random cnt))))
      (display (list 'next next))
      (it-n-or-last (+ next 1))
      (display 'redraw)
      (it-redraw)
      (display 'iter)
      (slideshow-proc (if n next n) cnt time)))
  (set! *slideshow*
        ;; this should run in a separate thread, but imlib_render_image_part_on_drawable_at_size hangs that way
        (slideshow-proc (if rand #f (p-get-file-index)) (p-get-file-count) time))
  (set! *waiter* (lambda (key ctrl mod1)
                   (cancel-thread *slideshow*)
                   (set! *waiter* default-waiter)
                   #t))
  #t)

(define (take-up-to lst k)
  (if (>= (length lst) k)
      (take lst k)
      lst))

(define (parse-page host page regexp)
  (define (load-page host page)
    (define* (read-string port #:optional (result ""))
      (let ((line (read-char port)))
        (if (eof-object? line)
            result
            (read-string port (string-append result (string line))))))
    (let* ((ai (car (getaddrinfo host "http")))
           (s  (socket (addrinfo:fam ai)
                       (addrinfo:socktype ai)
                       (addrinfo:protocol ai))))
      (connect s (addrinfo:addr ai))
      (display (string-append "GET /" page " HTTP/1.1\n"
                              "User-Agent: sxiv\n"
                              "Host: " host "\n"
                              "Accept: text/html\n"
                              "Connection: close\n\n")
               s)
      (read-string s)))
  (map (lambda (match) (match:substring match 1))
       (list-matches regexp
                     (load-page host page))))

(define (sort-of-url-encode str)
  (string-map (lambda (c) (if (eqv? c #\space)
                         #\+
                         c))
              str))

(define (load-imgur-images raw-query page max-galleries max-images-per-gallery max-images)
  (display "loading links") (newline)
  (let* ((query (sort-of-url-encode raw-query))
         (galleries (take-up-to (filter (lambda (gallery) (not (string=? gallery "random")))
                                        (parse-page "imgur.com"
                                                    (string-append "gallery/hot/viral/day/page/"
                                                                   (number->string page)
                                                                   "/hit?scrolled&set=0&q="
                                                                   query)
                                                    "href=\"/gallery/([^\"]+)\""))
                                max-galleries))
         (images (take-up-to (append-map (lambda (gallery) (take-up-to (parse-page "imgur.com"
                                                                              (string-append "gallery/" gallery)
                                                                              "img src=\"//i.imgur.com/([a-z0-9A-Z]+\\.[a-z]+)\"")
                                                                  max-images-per-gallery))
                                         galleries)
                             max-images))
         (directory (string-append "/tmp/" query "-" (number->string (random (* 42 42))) "/")))
    (display "links are loaded") (newline)
    (system (string-append "mkdir " directory))
    (map (lambda (image) (begin
                      (display (string-append "loading http://i.imgur.com/" image)) (newline)
                      (system (string-append "wget -qO " directory image " http://i.imgur.com/" image))
                      (it-add-image (string-append directory image))))
         images)
    #t))


(define (apply-input-to func)
  (display "waiting for an input")
  (newline)
  (set! *input* "")
  (set! *waiter*
        (lambda (key ctrl mod1)
          (let ((char (integer->char key)))
            (cond ((and ctrl
                        (eqv? (integer->char (+ key 96)) #\g))
                   (begin (set! *waiter* default-waiter)
                          #t))
                  ((eqv? char #\return) (begin
                                          (set! *waiter* default-waiter)
                                          (func *input*)
                                          #t))
                  ((eqv? char #\backspace) (begin (set! *input* (xsubstring *input*
                                                                            0
                                                                            (- (string-length *input*) 1)))
                                                  (p-set-bar-left *input*)
                                                  #t))
                  ((<= key 31) #f)
                  (else (begin (set! *input* (string-append *input* (string char)))
                               (p-set-bar-left *input*)
                               #t))))))
  (p-set-bar-left "")
  #t)

(define (apply-numeric-input-to func)
  (apply-input-to (lambda (numstr)
                    (func (if (string->number numstr)
                              (string->number numstr)
                              0)))))

(define (default-waiter key ctrl mod1)
  ;(newline)
  (display (list 'command key ctrl mod1))
  (newline)
  (if (> key 0)
      (if ctrl
          (match (integer->char (+ key 96))
            (#\6 (i-alternate))
            (#\n (i-navigate-frame 1))
            (#\p (i-navigate-frame -1))
            (#\m (i-toggle-animation))
            (#\h (it-scroll-screen (const left)))
            (#\j (it-scroll-screen (const down)))
            (#\k (it-scroll-screen (const up)))
            (#\l (it-scroll-screen (const right)))
            (#\e (apply-input-to (lambda (str) (p-set-bar-left (object->string (eval-string str))))))
            (#\a (apply-input-to it-add-image))
            (#\s (start-slideshow #f 2))
            (#\i (apply-input-to (lambda (query) (load-imgur-images query 0 15 5 50))))
            (else #f))
          (match (integer->char key)
            (#\q (it-quit))
            (#\return (it-switch-mode))
            (#\f (it-toggle-fullscreen))
            (#\b (it-toggle-bar))
            (#\r (it-reload-image))
            (#\R (t-reload-all))
            (#\D (it-remove-image))
            (#\n (i-navigate 1))
            (#\space (i-navigate 1))
            (#\p (i-navigate -1))
            (#\backspace (i-navigate -1))
            (#\g (it-first))
            (#\G (apply-numeric-input-to it-n-or-last))
            (#\h (apply-numeric-input-to (lambda (num) (it-scroll-move (const left) num))))
            (#\j (apply-numeric-input-to (lambda (num) (it-scroll-move (const down) num))))
            (#\k (apply-numeric-input-to (lambda (num) (it-scroll-move (const up) num))))
            (#\l (apply-numeric-input-to (lambda (num) (it-scroll-move (const right) num))))
            (#\H (i-scroll-to-edge (const left)))
            (#\J (i-scroll-to-edge (const down)))
            (#\K (i-scroll-to-edge (const up)))
            (#\L (i-scroll-to-edge (const right)))
            (#\+ (i-zoom 1))
            (#\- (i-zoom -1))
            (#\= (i-set-zoom 100))
            (#\w (i-fit-to-win (const scale-fit)))
            (#\e (i-fit-to-win (const scale-width)))
            (#\E (i-fit-to-win (const scale-height)))
            (#\W (i-fit-to-img))
            (#\< (i-rotate (const degree-270)))
            (#\> (i-rotate (const degree-90)))
            (#\? (i-rotate (const degree-180)))
            (#\| (i-flip (const flip-horizontal)))
            (#\_ (i-flip (const flip-vertical)))
            (#\a (i-toggle-antialias))
            (#\A (it-toggle-alpha))
            (else #f)))
      #f))

(define *waiter* default-waiter)

(define (on-key-press key ctrl mod1)
  (*waiter* key ctrl mod1))

(define (on-button-press button ctrl x y)
  (display (list button ctrl x y))
  (match button
    (4 (if ctrl
           (it-scroll-move (const left) 42)
           (it-scroll-move (const up) 42)))
    (5 (if ctrl
           (it-scroll-move (const right) 42)
           (it-scroll-move (const down) 42)))
    (2 (it-toggle-fullscreen))
    (1 (i-navigate 1))
    (3 (i-navigate -1))
    (else #f)))
